### PR TITLE
oauth2_proxy 2.0.1 (new formula)

### PR DESCRIPTION
**Pull request by ebardsley**

### CODE REVIEW

File: Library/Formula/oauth2_proxy.rb [hunk 1]
```diff
@@ -0,0 +1,80 @@
+require "language/go"
+
+class Oauth2Proxy < Formula
+  desc "A reverse proxy that provides authentication with Google, Github or other provider"
```
Comment (bfontaine) [inline Library/Formula/oauth2_proxy.rb]: Please use a shorter description. “`oauth2_proxy: <description>`” should fit under 80 characters.

File: Library/Formula/oauth2_proxy.rb [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+  def install
+    mkdir_p "#{buildpath}/src/github.com/bitly"
+    ln_s buildpath, "#{buildpath}/src/github.com/bitly/oauth2_proxy"
+
+    ENV["GOPATH"] = buildpath
+
+    Language::Go.stage_deps resources, buildpath/"src"
+    system "go", "build", "-o", "oauth2_proxy"
+    bin.install "oauth2_proxy"
```
Comment (bfontaine) [inline Library/Formula/oauth2_proxy.rb]: You can probably merge these two lines with `system "go", "build", "-o", "#{bin}/oauth2_proxy"`

File: Library/Formula/oauth2_proxy.rb [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    doc.install "README.md"
+
+    (etc/"oauth2_proxy").mkpath
+    cfg = etc/"oauth2_proxy"/"oauth2_proxy.cfg"
+    if !cfg.exist?
+      touch cfg
+    end
```
Comment (bfontaine) [inline Library/Formula/oauth2_proxy.rb]: You can do that in one line: `touch cfg unless cfg.exist?`

File: Library/Formula/oauth2_proxy.rb [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+  end
+
+  test do
+    system "#{bin}/oauth2_proxy", "--version"
+  end
+
+  def caveats; <<-EOS.undent
+    #{etc}/oauth2_proxy/oauth2_proxy.cfg must be filled in.
+    EOS
+  end
```
Comment (bfontaine) [inline Library/Formula/oauth2_proxy.rb]: Please move the caveats above the test block.

File: Library/Formula/oauth2_proxy.rb (hunk 3 above)
Comment (mikemcquaid) [inline Library/Formula/oauth2_proxy.rb]: Just use `etc.install` here instead which can create the directory and install an empty file.

File: Library/Formula/oauth2_proxy.rb (hunk 1 above)
Comment (ebardsley) [inline Library/Formula/oauth2_proxy.rb]: Done. (FWIW, there seemed to be a theme in the instructions about following precedent from other repositories, which is why I copied the github project 1-liner verbatim.)

File: Library/Formula/oauth2_proxy.rb (hunk 3 above)
Comment (ebardsley) [inline Library/Formula/oauth2_proxy.rb]: Changed to use install. Is there a better way to pull in an empty file? (It seemed ugly explicitly do something with /dev/null.)

**Comment by ebardsley** [comment]
Changes done and rebased in.

File: Library/Formula/oauth2_proxy.rb (hunk 3 above)
Comment (mikemcquaid) [inline Library/Formula/oauth2_proxy.rb]: Your way looks nice :+1:

File: Library/Formula/oauth2_proxy.rb [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    Language::Go.stage_deps resources, buildpath/"src"
+    system "go", "build", "-o", "#{bin}/oauth2_proxy"
+    doc.install "README.md"
+
+    touch "oauth2_proxy.cfg"
+    (etc/"oauth2_proxy").install "oauth2_proxy.cfg"
+  end
+
+  def caveats; <<-EOS.undent
+    #{etc}/oauth2_proxy/oauth2_proxy.cfg must be filled in.
```
Comment (DomT4) [inline Library/Formula/oauth2_proxy.rb]: Can we write a basic configuration file for them?

**Comment by ebardsley** [comment]
@DomT4 , done. (I forgot that there was a contributed example config.)

File: Library/Formula/oauth2_proxy.rb [hunk 6]
```diff
@@ -0,0 +1,74 @@
...
+    (etc/"oauth2_proxy").install "contrib/oauth2_proxy.cfg.example"
+  end
+
+  def caveats; <<-EOS.undent
+    #{etc}/oauth2_proxy/oauth2_proxy.cfg must be filled in.
+    EOS
+  end
+
+  test do
+    system "#{bin}/oauth2_proxy", "--version"
```
Comment (mikemcquaid) [inline Library/Formula/oauth2_proxy.rb]: Can this test be modified to do something more substantial than e.g. `--version` or `--help`? See `cmake.rb` for an example of an application formula with a good test and `tinyxml2.rb` for an example of a library formula with a good test. Thanks!

Comment (ebardsley) [inline Library/Formula/oauth2_proxy.rb]: I don't think so. It's a long-running daemon and doesn't have a
--validate-config-and-exit style option at the moment.

File: Library/Formula/oauth2_proxy.rb [hunk 7]
```diff
@@ -0,0 +1,74 @@
...
+      <dict>
+        <key>Label</key>
+        <string>#{plist_name}</string>
+        <key>RunAtLoad</key>
+        <true/>
+        <key>KeepAlive</key>
+        <true/>
+        <key>ProgramArguments</key>
+        <array>
+            <string>#{bin}/oauth2_proxy</string>
```
Comment (DomT4) [inline Library/Formula/oauth2_proxy.rb]: This should be `opt_bin` so it doesn't break across updates.

File: Library/Formula/oauth2_proxy.rb [hunk 8]
```diff
@@ -0,0 +1,74 @@
+require "language/go"
+
+class Oauth2Proxy < Formula
+  desc "reverse proxy for authenticating users via OAuth 2 providers"
```
Comment (DomT4) [inline Library/Formula/oauth2_proxy.rb]: Start with a capital letter please.

File: Library/Formula/oauth2_proxy.rb (hunk 6 above)
Comment (mikemcquaid) [inline Library/Formula/oauth2_proxy.rb]: You can use `fork do` and test the forked process.

**Comment by ebardsley** [comment]
@mikemcquaid I've added an end-to-end test; let me know what you think.

@DomT4 Changes made.

File: Library/Formula/oauth2_proxy.rb [hunk 9]
```diff
@@ -0,0 +1,102 @@
...
+    server = TCPServer.new(0)
+    port = server.addr[1]
+    server.close
+
+    pid = fork do
+      exec "#{bin}/oauth2_proxy",
+        "--client-id=testing",
+        "--client-secret=testing",
+        "--cookie-secret=testing",
+        "--http-address=127.0.0.1:" + port.to_s,
```
Comment (mikemcquaid) [inline Library/Formula/oauth2_proxy.rb]: `"--http-address=127.0.0.1:#{port}"`

Comment (mikemcquaid) [inline Library/Formula/oauth2_proxy.rb]: I'd just pick a random port here though; I don't think you need to use a TCPServer as this is primary for just using on the bots.

**Comment by mikemcquaid** [comment]
Nice test :+1:

**Comment by ebardsley** [comment]
@mikemcquaid Changed to use string interpolation. I'd rather keep this as a unused port, as it makes the test more robust on a wide variety of systems (eg, someone else has the same "favorite random port" as me) and it also makes it safer to run two copies of this test concurrently.

**Comment by mikemcquaid** [comment]
Thanks for your contribution to Homebrew! Without people like you submitting PRs we couldn't run this project. You rock!

**Pull request closed without merging**